eth2/beacon/chains/base.py

```diff
@@ -0,0 +1,406 @@
+                    type(genesis_block),
+                    sm_class.block_class
+                )
+            )
+
+        chaindb = cls.get_chaindb_class()(db=base_db)
+        chaindb.persist_state(genesis_state)
+        return cls.from_genesis_block(base_db, genesis_block)
+
+    @classmethod
+    def from_genesis_block(cls,
+                           base_db: BaseAtomicDB,
+                           genesis_block: BaseBeaconBlock) -> 'BaseBeaconChain':
+        """
+        Initialize the ``BeaconChain`` from the genesis block.
+        """
+        chaindb = cls.get_chaindb_class()(db=base_db)
+        chaindb.persist_block(genesis_block, genesis_block.__class__)
+        return cls(base_db)
+
+    #
+    # StateMachine API
+    #
+    @classmethod
+    def get_state_machine_class(cls, block: BaseBeaconBlock) -> Type['BaseBeaconStateMachine']:
+        """
+        Returns the ``StateMachine`` instance for the given block slot number.
+        """
+        return cls.get_state_machine_class_for_block_slot(block.slot)
+
+    @classmethod
+    def get_state_machine_class_for_block_slot(
+            cls,
+            slot: SlotNumber) -> Type['BaseBeaconStateMachine']:
+        """
+        Return the ``StateMachine`` class for the given block slot number.
+        """
+        if cls.sm_configuration is None:
+            raise AttributeError("Chain classes must define the StateMachines in sm_configuration")
+
+        validate_slot(slot)
+        for start_slot, sm_class in reversed(cls.sm_configuration):
+            if slot >= start_slot:
+                return sm_class
+        raise StateMachineNotFound("No StateMachine available for block slot: #{0}".format(slot))
+
+    def get_state_machine(self, at_block: BaseBeaconBlock=None) -> 'BaseBeaconStateMachine':
+        """
+        Return the ``StateMachine`` instance for the given block number.
+        """
+        block = self.ensure_block(at_block)
+        sm_class = self.get_state_machine_class_for_block_slot(block.slot)
+        parent_block_class = self.get_block_class(block.parent_root)
+        return sm_class(
+            chaindb=self.chaindb,
+            block=block,
+            parent_block_class=parent_block_class,
+        )
+
+    #
+    # Block API
+    #
+    def get_block_class(self, block_root: Hash32) -> Type[BaseBeaconBlock]:
+        slot = self.chaindb.get_slot_by_root(block_root)
+        sm_class = self.get_state_machine_class_for_block_slot(slot)
+        block_class = sm_class.block_class
+        return block_class
+
+    def create_block_from_parent(self,
+                                 parent_block: BaseBeaconBlock,
+                                 block_params: FromBlockParams) -> BaseBeaconBlock:
+        """
+        Passthrough helper to the ``StateMachine`` class of the block descending from the
+        given block.
+        """
+
+        return self.get_state_machine_class_for_block_slot(
+            slot=parent_block.slot + 1,
```

Comment:
maybe
```suggestion
            slot=parent_block.slot + missed_slots + 1,
```
where `missed_slots` is an additional parameter with default `0`?